src/components/crossDevice/CrossDeviceLink/index.js

```diff
@@ -307,18 +327,48 @@ class CrossDeviceLinkUI extends Component {
     this.viewOptionBtn.blur()
   }
 
+  getRequiredViewRenders = () => {
+    const { _crossDeviceLinkMethods } = this.props
+    const defaultViewRendersMap = {
+      qr_code: this.renderQrCodeSection,
+      sms: this.renderSmsLinkSection,
+      copy_link: this.renderCopyLinkSection,
+    }
+    if (
+      !_crossDeviceLinkMethods ||
+      (_crossDeviceLinkMethods && _crossDeviceLinkMethods.length < 1)
+    ) {
+      return defaultViewRendersMap
+    }
+    if (
+      _crossDeviceLinkMethods &&
+      configHasInvalidViewIds(_crossDeviceLinkMethods)
+    ) {
+      return defaultViewRendersMap
+    }
+
+    const requiredViewRendersMap = _crossDeviceLinkMethods.reduce(
+      (result, viewId) => {
+        result[viewId] = defaultViewRendersMap[viewId]
+        return result
+      },
+      {}
+    )
+    return requiredViewRendersMap
+  }
+
   componentWillUnmount() {
     this.clearSendLinkClickTimeout()
   }
 
   render() {
-    const { translate, trackScreen } = this.props
+    const { translate, trackScreen, _crossDeviceLinkMethods } = this.props
     const { error, currentViewId } = this.state
-    const currentViewRender = {
-      qr_code: this.renderQrCodeSection,
-      sms: this.renderSmsLinkSection,
-      copy_link: this.renderCopyLinkSection,
-    }[currentViewId]
+    const requiredViewRenders = this.getRequiredViewRenders()
+    const currentViewRender = requiredViewRenders[currentViewId]
+    const visibleViewOptions = SECURE_LINK_VIEWS.filter(
+      (view) => (view = Object.keys(requiredViewRenders).includes(view.id))
+    )
```

Comment:
💅🏻 [nitpick]

I don't think `view =` is doing anything, is it?
Also, we can use keyword `in` to check whether the object has that key defined and reduce time complexity.

```suggestion
    const visibleViewOptions = SECURE_LINK_VIEWS.filter(
      (view) => view.id in requiredViewRenders
    )
```

Reply:
Interesting. I've never used that approach with the `in` keyword for that check. And yes, `view =` isn't required there. Might have got left there following a rebase and/or me cleaning up the code.

Reply:
It's almost the same thing as using a Set too, like `new Set(Object.keys(requiredViewRenders))`.
Except plain objects already are indexed by their keys so we don't need the Set.